File: python/src/graph_engineering/typed_ports.py
"""Opt-in strict-exact typed-port validation for Graph IR v1alpha1."""

from __future__ import annotations

from typing import TypeAlias

from jsonschema import Draft202012Validator  # type: ignore[import-untyped]
from jsonschema.exceptions import SchemaError  # type: ignore[import-untyped]

from .canonical import canonical_json
from .compiler import Diagnostic, DiagnosticCode
from .models import GraphSpec, JsonObject, JsonValue, NodeSpec

TYPED_PORT_POLICY_KEY = "graphengineering.reacher-z.github.io/typed-ports"
TYPED_PORT_POLICY_API_VERSION = "graphengineering.reacher-z.github.io/typed-ports/v1alpha1"
TYPED_PORT_POLICY_MODE = "strict-exact"
TYPED_PORT_POLICY_PATH = "#/policies/graphengineering.reacher-z.github.io~1typed-ports"
TYPED_PORT_SCHEMA_DIALECT = "https://json-schema.org/draft/2020-12/schema"

SchemaValue: TypeAlias = JsonObject | bool


def _pointer(path: str, part: str | int) -> str:
    escaped = str(part).replace("~", "~0").replace("/", "~1")
    return f"{path}/{escaped}" if path != "#" else f"#/{escaped}"


def _uses_unsupported_schema_feature(schema: JsonObject) -> bool:
    remaining: list[JsonObject] = [schema]
    while remaining:
        current = remaining.pop()
        if "$ref" in current or "$dynamicRef" in current:
            return True
        if "pattern" in current or "patternProperties" in current:
            return True
        if "enum" in current and type(current["enum"]) is list and not current["enum"]:
            return True
        if "$schema" in current and current["$schema"] != TYPED_PORT_SCHEMA_DIALECT:
            return True

        # Walk only locations whose JSON Schema vocabulary declares another
        # schema. Arbitrary data below const/default/examples/enum, and names
        # inside schema maps, are not themselves schemas and must not be
        # mistaken for unsupported keywords.
        for keyword in ("$defs", "definitions", "properties", "dependentSchemas"):
            schema_map = current.get(keyword)
            if isinstance(schema_map, dict):
                remaining.extend(value for value in schema_map.values() if isinstance(value, dict))

        dependencies = current.get("dependencies")
        if isinstance(dependencies, dict):
            remaining.extend(value for value in dependencies.values() if isinstance(value, dict))

        for keyword in (
            "additionalProperties",
            "unevaluatedProperties",
            "propertyNames",
            "contains",
            "not",
            "if",
            "then",
            "else",
            "items",
            "unevaluatedItems",
            "contentSchema",
        ):
            nested = current.get(keyword)
            if isinstance(nested, dict):
                remaining.append(nested)

        for keyword in ("prefixItems", "allOf", "anyOf", "oneOf"):
            nested_array = current.get(keyword)
            if isinstance(nested_array, list):
                remaining.extend(value for value in nested_array if isinstance(value, dict))
    return False


def _schema_is_valid(schema: JsonObject) -> bool:
    if _uses_unsupported_schema_feature(schema):
        return False
    try:
        Draft202012Validator.check_schema(schema)
    except (SchemaError, RecursionError):
        return False
    return True


def _schema_diagnostic(
    path: str,
    *,
    node_id: str | None = None,
    node_ids: tuple[str, ...] | None = None,
    edge_id: str | None = None,
) -> Diagnostic:
    return Diagnostic(
        code=DiagnosticCode.INVALID_PORT_SCHEMA,
        message="strict typed-port schema is not supported by the v1alpha1 profile",
        node_id=node_id,
        node_ids=node_ids,
        edge_id=edge_id,
        path=path,
    )


def _node_ids(*values: str) -> tuple[str, ...]:
    """Keep declaration-role order while projecting identifier set semantics."""

    return tuple(dict.fromkeys(values))


def _policy_value(graph: GraphSpec) -> JsonValue | None:
    if graph.policies is None:
        return None
    return (graph.policies.model_extra or {}).get(TYPED_PORT_POLICY_KEY)


def _policy_enabled(graph: GraphSpec) -> tuple[bool, tuple[Diagnostic, ...]]:
    value = _policy_value(graph)
    if graph.policies is None or TYPED_PORT_POLICY_KEY not in (graph.policies.model_extra or {}):
        return False, ()
    expected: JsonObject = {
        "apiVersion": TYPED_PORT_POLICY_API_VERSION,
        "mode": TYPED_PORT_POLICY_MODE,
    }
    if type(value) is not dict or value != expected:
        return (
            True,
            (
                Diagnostic(
                    code=DiagnosticCode.INVALID_PORT_SCHEMA,
                    message="strict typed-port policy must match the v1alpha1 profile exactly",
                    path=TYPED_PORT_POLICY_PATH,
                ),
            ),
        )
    return True, ()


def _schema_profile_diagnostics(graph: GraphSpec) -> tuple[Diagnostic, ...]:
    checks: list[tuple[JsonObject, str, str | None, tuple[str, ...] | None, str | None]] = [
        (graph.input_schema, "#/inputSchema", None, None, None),
        (graph.output_schema, "#/outputSchema", None, None, None),
    ]
    if graph.state_schema is not None:
        checks.append((graph.state_schema, "#/stateSchema", None, None, None))
    for index, node in enumerate(graph.nodes):
        checks.extend(
            (
                (
                    node.input_schema,
                    f"#/nodes/{index}/inputSchema",
                    node.id,
                    (node.id,),
                    None,
                ),
                (
                    node.output_schema,
                    f"#/nodes/{index}/outputSchema",
                    node.id,
                    (node.id,),
                    None,
                ),
            )
        )
    for index, edge in enumerate(graph.edges):
        if edge.schema_ is not None:
            checks.append(
                (
                    edge.schema_,
                    f"#/edges/{index}/schema",
                    None,
                    None,
                    edge.id,
                )
            )

    diagnostics: list[Diagnostic] = []
    for schema, path, node_id, node_ids, edge_id in checks:
        if not _schema_is_valid(schema):
            diagnostics.append(
                _schema_diagnostic(
                    path,
                    node_id=node_id,
                    node_ids=node_ids,
                    edge_id=edge_id,
                )
            )
    return tuple(diagnostics)


def _object_property(schema: JsonObject, name: str) -> SchemaValue | None:
    if schema.get("type") != "object":
        return None
    properties = schema.get("properties")
    required = schema.get("required")
    if not isinstance(properties, dict) or not isinstance(required, list):
        return None
    if name not in required:
        return None
    value = properties.get(name)
    if type(value) is dict:
        return value
    if type(value) is bool:
        return value
    return None


def _schemas_equal(left: SchemaValue, right: SchemaValue) -> bool:
    return canonical_json(left) == canonical_json(right)


def _edge_binding_diagnostics(
    graph: GraphSpec,
    node_by_id: dict[str, NodeSpec],
) -> list[Diagnostic]:
    diagnostics: list[Diagnostic] = []
    seen_bindings: set[tuple[str, str]] = set()

    for index, edge in enumerate(graph.edges):
        edge_path = f"#/edges/{index}"
        producer = node_by_id[edge.source.node]
        consumer = node_by_id[edge.target.node]
        if edge.mode not in (None, "value"):
            diagnostics.append(
                Diagnostic(
                    code=DiagnosticCode.UNSUPPORTED_TYPED_EDGE_MODE,
                    message="strict typed ports support only value edges",
                    node_id=producer.id,
                    node_ids=_node_ids(producer.id, consumer.id),
                    edge_id=edge.id,
                    path=f"{edge_path}/mode",
                )
            )

        source_schema: SchemaValue | None
        if edge.source.port is None:
            source_schema = producer.output_schema
        else:
            source_schema = _object_property(producer.output_schema, edge.source.port)
            if source_schema is None:
                diagnostics.append(
                    Diagnostic(
                        code=DiagnosticCode.MISSING_SOURCE_PORT,
                        message="source port is not a required output-schema property",
                        node_id=producer.id,
                        node_ids=(producer.id,),
                        edge_id=edge.id,
                        path=f"{edge_path}/from/port",
                    )
                )

        binding_key = edge.target.port or edge.source.node
        target_schema = _object_property(consumer.input_schema, binding_key)
        if target_schema is None:
            diagnostics.append(
                Diagnostic(
                    code=DiagnosticCode.MISSING_TARGET_PORT,
                    message="target binding is not a required input-schema property",
                    node_id=consumer.id,
                    node_ids=(consumer.id,),
                    edge_id=edge.id,
                    path=f"{edge_path}/to" + ("/port" if edge.target.port is not None else ""),
                )
            )

        binding = (consumer.id, binding_key)
        if binding in seen_bindings:
            diagnostics.append(
                Diagnostic(
                    code=DiagnosticCode.DUPLICATE_TARGET_BINDING,
                    message="multiple edges bind the same target input property",
                    node_id=consumer.id,
                    node_ids=(consumer.id,),
                    edge_id=edge.id,
                    path=f"{edge_path}/to" + ("/port" if edge.target.port is not None else ""),
                )
            )
        else:
            seen_bindings.add(binding)

        if source_schema is None or target_schema is None:
            continue
        schemas = [source_schema, target_schema]
        if edge.schema_ is not None:
            schemas.append(edge.schema_)
        if any(not _schemas_equal(schemas[0], item) for item in schemas[1:]):
            diagnostics.append(
                Diagnostic(
                    code=DiagnosticCode.PORT_SCHEMA_MISMATCH,
                    message="source, edge, and target schemas must be canonical-identical",
                    node_id=producer.id,
                    node_ids=_node_ids(producer.id, consumer.id),
                    edge_id=edge.id,
                    path=edge_path,
                )
            )
    return diagnostics


def _output_diagnostics(
    graph: GraphSpec,
    node_by_id: dict[str, NodeSpec],
) -> list[Diagnostic]:
    diagnostics: list[Diagnostic] = []
    for output_name in sorted(graph.outputs):
        endpoint = graph.outputs[output_name]
        node = node_by_id[endpoint.node]
        output_path = _pointer("#/outputs", output_name)
        selected_schema: SchemaValue | None
        if endpoint.port is None:
            selected_schema = node.output_schema
        else:
            selected_schema = _object_property(node.output_schema, endpoint.port)
            if selected_schema is None:
                diagnostics.append(
                    Diagnostic(
                        code=DiagnosticCode.MISSING_SOURCE_PORT,
                        message="public output port is not a required output-schema property",
                        node_id=node.id,
                        node_ids=(node.id,),
                        output_name=output_name,
                        path=f"{output_path}/port",
                    )
                )

        public_schema = _object_property(graph.output_schema, output_name)
        if public_schema is None:
            diagnostics.append(
                Diagnostic(
                    code=DiagnosticCode.OUTPUT_SCHEMA_MISMATCH,
                    message="public output is not a required graph output-schema property",
                    node_id=node.id,
                    node_ids=(node.id,),
                    output_name=output_name,
                    path=output_path,
                )
            )
        elif selected_schema is not None and not _schemas_equal(selected_schema, public_schema):
            diagnostics.append(
                Diagnostic(
                    code=DiagnosticCode.OUTPUT_SCHEMA_MISMATCH,
                    message="public output binding schema does not match graph outputSchema",
                    node_id=node.id,
                    node_ids=(node.id,),
                    output_name=output_name,
                    path=output_path,
                )
            )
    return diagnostics


def validate_strict_typed_ports(graph: GraphSpec) -> tuple[Diagnostic, ...]:
    """Return deterministic strict-exact diagnostics for an opted-in graph."""

    enabled, policy_diagnostics = _policy_enabled(graph)
    if not enabled:
        return ()
    if policy_diagnostics:
        return policy_diagnostics

    schema_diagnostics = _schema_profile_diagnostics(graph)
    if schema_diagnostics:
        return schema_diagnostics

    node_by_id = {node.id: node for node in graph.nodes}
    diagnostics: list[Diagnostic] = []
    for index, entrypoint in enumerate(graph.entrypoints):
        node = node_by_id[entrypoint]
        if not _schemas_equal(graph.input_schema, node.input_schema):
            diagnostics.append(
                Diagnostic(
                    code=DiagnosticCode.ENTRYPOINT_SCHEMA_MISMATCH,
                    message="entrypoint input schema must match graph inputSchema exactly",
                    node_id=node.id,
                    node_ids=(node.id,),
                    path=f"#/entrypoints/{index}",
                )
            )

    diagnostics.extend(_edge_binding_diagnostics(graph, node_by_id))
    diagnostics.extend(_output_diagnostics(graph, node_by_id))
    return tuple(diagnostics)


def strict_typed_ports_policy() -> JsonObject:
    """Return a fresh portable policy extension value for builder helpers."""

    return {
        "apiVersion": TYPED_PORT_POLICY_API_VERSION,
        "mode": TYPED_PORT_POLICY_MODE,
    }


__all__ = [
    "TYPED_PORT_POLICY_API_VERSION",
    "TYPED_PORT_POLICY_KEY",
    "TYPED_PORT_POLICY_MODE",
    "TYPED_PORT_SCHEMA_DIALECT",
    "strict_typed_ports_policy",
    "validate_strict_typed_ports",
]

File: packages/core/src/typed-ports.ts
import { Ajv2020 } from "ajv/dist/2020.js";
import {
  canonicalSerialize,
  captureCanonicalJson,
  compareUnicodeCodePoints,
} from "./canonical.js";
import type { CompilerDiagnostic } from "./compiler.js";
import type { EdgeSpec, GraphSpec, JsonSchema } from "./types.js";

export const STRICT_TYPED_PORTS_POLICY_KEY =
  "graphengineering.reacher-z.github.io/typed-ports" as const;
export const STRICT_TYPED_PORTS_API_VERSION =
  "graphengineering.reacher-z.github.io/typed-ports/v1alpha1" as const;
export const STRICT_TYPED_PORTS_MODE = "strict-exact" as const;

const DRAFT_2020_12_DIALECT =
  "https://json-schema.org/draft/2020-12/schema" as const;

export interface StrictTypedPortsPolicy {
  readonly apiVersion: typeof STRICT_TYPED_PORTS_API_VERSION;
  readonly mode: typeof STRICT_TYPED_PORTS_MODE;
}

const POLICY_PATH =
  "#/policies/graphengineering.reacher-z.github.io~1typed-ports";

const schemaValidator = new Ajv2020({
  addUsedSchema: false,
  allErrors: true,
  logger: false,
  strict: false,
  validateFormats: false,
});

function record(value: unknown): value is Readonly<Record<string, unknown>> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

function pointerSegment(value: string): string {
  return value.replaceAll("~", "~0").replaceAll("/", "~1");
}

function diagnostic(
  code:
    | "GE1201_MISSING_SOURCE_PORT"
    | "GE1202_MISSING_TARGET_PORT"
    | "GE1203_PORT_SCHEMA_MISMATCH"
    | "GE1204_DUPLICATE_TARGET_BINDING"
    | "GE1205_INVALID_PORT_SCHEMA"
    | "GE1206_OUTPUT_SCHEMA_MISMATCH"
    | "GE1207_ENTRYPOINT_SCHEMA_MISMATCH"
    | "GE1208_UNSUPPORTED_TYPED_EDGE_MODE",
  message: string,
  path: string,
  fields: Pick<CompilerDiagnostic, "nodeIds" | "edgeId" | "outputName"> = {},
): CompilerDiagnostic {
  const nodeIds = Object.freeze([...(fields.nodeIds ?? [])]);
  return Object.freeze({
    code,
    severity: "error",
    message,
    path,
    nodeIds,
    ...(fields.edgeId === undefined ? {} : { edgeId: fields.edgeId }),
    ...(fields.outputName === undefined ? {} : { outputName: fields.outputName }),
  });
}

function schemaEquals(left: unknown, right: unknown): boolean {
  return canonicalSerialize(left) === canonicalSerialize(right);
}

const SCHEMA_MAP_KEYWORDS = [
  "$defs",
  "definitions",
  "properties",
  "dependentSchemas",
] as const;
const SCHEMA_VALUE_KEYWORDS = [
  "additionalProperties",
  "unevaluatedProperties",
  "propertyNames",
  "contains",
  "not",
  "if",
  "then",
  "else",
  "items",
  "unevaluatedItems",
  "contentSchema",
] as const;
const SCHEMA_ARRAY_KEYWORDS = ["prefixItems", "allOf", "anyOf", "oneOf"] as const;

function unsupportedSchemaProfile(value: unknown): boolean {
  if (typeof value === "boolean") return false;
  if (!record(value)) return false;
  if (
    Object.hasOwn(value, "$ref") ||
    Object.hasOwn(value, "$dynamicRef") ||
    Object.hasOwn(value, "pattern") ||
    Object.hasOwn(value, "patternProperties")
  ) {
    return true;
  }
  if (Object.hasOwn(value, "$schema") && value.$schema !== DRAFT_2020_12_DIALECT) return true;
  if (Array.isArray(value.enum) && value.enum.length === 0) return true;

  for (const keyword of SCHEMA_MAP_KEYWORDS) {
    const map = value[keyword];
    if (!record(map)) continue;
    for (const child of Object.values(map)) {
      if (unsupportedSchemaProfile(child)) return true;
    }
  }
  for (const keyword of SCHEMA_VALUE_KEYWORDS) {
    const child = value[keyword];
    if (Array.isArray(child)) {
      if (child.some(unsupportedSchemaProfile)) return true;
    } else if (unsupportedSchemaProfile(child)) {
      return true;
    }
  }
  for (const keyword of SCHEMA_ARRAY_KEYWORDS) {
    const children = value[keyword];
    if (Array.isArray(children) && children.some(unsupportedSchemaProfile)) return true;
  }
  const dependencies = value.dependencies;
  if (record(dependencies)) {
    for (const child of Object.values(dependencies)) {
      if ((record(child) || typeof child === "boolean") && unsupportedSchemaProfile(child)) {
        return true;
      }
    }
  }
  return false;
}

function validDraft202012Schema(value: unknown): boolean {
  // v1alpha1 intentionally refuses even local references. Comparing equal ref
  // tokens without resolving each independent schema root can manufacture a
  // false proof when the referenced definitions differ.
  if (unsupportedSchemaProfile(value)) return false;
  try {
    if (schemaValidator.validateSchema(value as object | boolean) !== true) return false;
    return true;
  } catch {
    return false;
  }
}

interface LocatedSchema {
  readonly value: unknown;
  readonly path: string;
  readonly nodeIds?: readonly string[];
  readonly edgeId?: string;
}

function participatingSchemas(graph: GraphSpec): readonly LocatedSchema[] {
  const schemas: LocatedSchema[] = [
    { value: graph.inputSchema, path: "#/inputSchema" },
    { value: graph.outputSchema, path: "#/outputSchema" },
  ];
  if (graph.stateSchema !== undefined) {
    schemas.push({ value: graph.stateSchema, path: "#/stateSchema" });
  }
  for (const [index, node] of graph.nodes.entries()) {
    schemas.push(
      { value: node.inputSchema, path: `#/nodes/${index}/inputSchema`, nodeIds: [node.id] },
      { value: node.outputSchema, path: `#/nodes/${index}/outputSchema`, nodeIds: [node.id] },
    );
  }
  for (const [index, edge] of graph.edges.entries()) {
    if (edge.schema !== undefined) {
      schemas.push({
        value: edge.schema,
        path: `#/edges/${index}/schema`,
        edgeId: edge.id,
      });
    }
  }
  return schemas;
}

type PortSelection =
  | { readonly found: false }
  | { readonly found: true; readonly schema: unknown };

function selectRequiredObjectProperty(schema: unknown, port: string): PortSelection {
  if (!record(schema) || schema.type !== "object") return { found: false };
  const properties = schema.properties;
  const required = schema.required;
  if (!record(properties) || !Object.hasOwn(properties, port)) return { found: false };
  if (!Array.isArray(required) || !required.includes(port)) return { found: false };
  return { found: true, schema: properties[port] };
}

function strictPolicyDiagnostic(graph: GraphSpec): CompilerDiagnostic | null {
  const policies = graph.policies;
  if (policies === undefined || !Object.hasOwn(policies, STRICT_TYPED_PORTS_POLICY_KEY)) {
    return null;
  }
  const value = policies[STRICT_TYPED_PORTS_POLICY_KEY];
  if (
    !record(value) ||
    Object.keys(value).length !== 2 ||
    value.apiVersion !== STRICT_TYPED_PORTS_API_VERSION ||
    value.mode !== STRICT_TYPED_PORTS_MODE
  ) {
    return diagnostic(
      "GE1205_INVALID_PORT_SCHEMA",
      "The strict typed-port policy must exactly select v1alpha1 strict-exact mode",
      POLICY_PATH,
    );
  }
  return null;
}

function hasStrictTypedPortsSnapshot(graph: GraphSpec): boolean {
  const policies = graph.policies;
  if (policies === undefined || !Object.hasOwn(policies, STRICT_TYPED_PORTS_POLICY_KEY)) {
    return false;
  }
  return strictPolicyDiagnostic(graph) === null;
}

function captureTypedGraph(graph: GraphSpec | unknown): GraphSpec | null {
  try {
    const captured = captureCanonicalJson(graph).value;
    return record(captured) ? captured as unknown as GraphSpec : null;
  } catch {
    return null;
  }
}

/** True only for the exact policy on a safely detached graph snapshot. */
export function hasStrictTypedPorts(graph: GraphSpec | unknown): boolean {
  const captured = captureTypedGraph(graph);
  return captured !== null && hasStrictTypedPortsSnapshot(captured);
}

function transferSchema(
  edge: EdgeSpec,
  edgeIndex: number,
  source: GraphSpec["nodes"][number],
  diagnostics: CompilerDiagnostic[],
): PortSelection {
  if (edge.from.port === undefined) return { found: true, schema: source.outputSchema };
  const selected = selectRequiredObjectProperty(source.outputSchema, edge.from.port);
  if (!selected.found) {
    diagnostics.push(
      diagnostic(
        "GE1201_MISSING_SOURCE_PORT",
        `Edge '${edge.id}' source port '${edge.from.port}' is not a required output property`,
        `#/edges/${edgeIndex}/from/port`,
        { edgeId: edge.id, nodeIds: [source.id] },
      ),
    );
  }
  return selected;
}

/**
 * Validate the conservative v1alpha1 strict-exact profile. Legacy graphs which
 * do not opt in return no diagnostics and make no static typed-port claim.
 */
export function validateStrictTypedPortsSnapshot(graph: GraphSpec): readonly CompilerDiagnostic[] {
  const policies = graph.policies;
  if (policies === undefined || !Object.hasOwn(policies, STRICT_TYPED_PORTS_POLICY_KEY)) {
    return Object.freeze([]);
  }

  const malformedPolicy = strictPolicyDiagnostic(graph);
  if (malformedPolicy !== null) return Object.freeze([malformedPolicy]);

  const profileDiagnostics: CompilerDiagnostic[] = [];
  for (const schema of participatingSchemas(graph)) {
    if (!validDraft202012Schema(schema.value)) {
      profileDiagnostics.push(
        diagnostic(
          "GE1205_INVALID_PORT_SCHEMA",
          "Schema is not valid Draft 2020-12 JSON Schema or contains an unsupported reference",
          schema.path,
          {
            ...(schema.nodeIds === undefined ? {} : { nodeIds: schema.nodeIds }),
            ...(schema.edgeId === undefined ? {} : { edgeId: schema.edgeId }),
          },
        ),
      );
    }
  }
  if (profileDiagnostics.length > 0) return Object.freeze(profileDiagnostics);

  const diagnostics: CompilerDiagnostic[] = [];
  const nodesById = new Map(graph.nodes.map((node) => [node.id, node]));

  for (const [index, entrypoint] of graph.entrypoints.entries()) {
    const node = nodesById.get(entrypoint);
    if (node !== undefined && !schemaEquals(graph.inputSchema, node.inputSchema)) {
      diagnostics.push(
        diagnostic(
          "GE1207_ENTRYPOINT_SCHEMA_MISMATCH",
          `Entrypoint '${entrypoint}' input schema does not exactly match graph inputSchema`,
          `#/entrypoints/${index}`,
          { nodeIds: [entrypoint] },
        ),
      );
    }
  }

  const targetBindings = new Map<string, string>();
  for (const [index, edge] of graph.edges.entries()) {
    const source = nodesById.get(edge.from.node);
    const target = nodesById.get(edge.to.node);
    if (source === undefined || target === undefined) continue;

    if (edge.mode !== undefined && edge.mode !== "value") {
      diagnostics.push(
        diagnostic(
          "GE1208_UNSUPPORTED_TYPED_EDGE_MODE",
          `Edge '${edge.id}' mode '${edge.mode}' is not supported by strict-exact typed ports`,
          `#/edges/${index}/mode`,
          { edgeId: edge.id, nodeIds: [source.id, target.id] },
        ),
      );
    }

    const sourceSelection = transferSchema(edge, index, source, diagnostics);
    const binding = edge.to.port ?? edge.from.node;
    const targetSelection = selectRequiredObjectProperty(target.inputSchema, binding);
    if (!targetSelection.found) {
      diagnostics.push(
        diagnostic(
          "GE1202_MISSING_TARGET_PORT",
          `Edge '${edge.id}' target binding '${binding}' is not a required input property`,
          edge.to.port === undefined ? `#/edges/${index}/to` : `#/edges/${index}/to/port`,
          { edgeId: edge.id, nodeIds: [target.id] },
        ),
      );
    }

    const targetBindingKey = `${target.id}\0${binding}`;
    const priorEdge = targetBindings.get(targetBindingKey);
    if (priorEdge === undefined) {
      targetBindings.set(targetBindingKey, edge.id);
    } else {
      diagnostics.push(
        diagnostic(
          "GE1204_DUPLICATE_TARGET_BINDING",
          `Edges '${priorEdge}' and '${edge.id}' write target binding '${binding}'`,
          edge.to.port === undefined ? `#/edges/${index}/to` : `#/edges/${index}/to/port`,
          { edgeId: edge.id, nodeIds: [target.id] },
        ),
      );
    }

    if (sourceSelection.found && targetSelection.found) {
      const edgeMatches =
        edge.schema === undefined || schemaEquals(sourceSelection.schema, edge.schema);
      if (!schemaEquals(sourceSelection.schema, targetSelection.schema) || !edgeMatches) {
        diagnostics.push(
          diagnostic(
            "GE1203_PORT_SCHEMA_MISMATCH",
            `Edge '${edge.id}' source, edge, and target schemas are not canonical-identical`,
            `#/edges/${index}`,
            { edgeId: edge.id, nodeIds: [source.id, target.id] },
          ),
        );
      }
    }
  }

  for (const outputName of Object.keys(graph.outputs).sort(compareUnicodeCodePoints)) {
    const endpoint = graph.outputs[outputName];
    if (endpoint === undefined) continue;
    const node = nodesById.get(endpoint.node);
    if (node === undefined) continue;

    const graphSelection = selectRequiredObjectProperty(graph.outputSchema, outputName);
    if (!graphSelection.found) {
      diagnostics.push(
        diagnostic(
          "GE1206_OUTPUT_SCHEMA_MISMATCH",
          `Public output '${outputName}' is not a required graph outputSchema property`,
          `#/outputs/${pointerSegment(outputName)}`,
          { outputName, nodeIds: [node.id] },
        ),
      );
      continue;
    }

    let nodeSelection: PortSelection;
    if (endpoint.port === undefined) {
      nodeSelection = { found: true, schema: node.outputSchema };
    } else {
      nodeSelection = selectRequiredObjectProperty(node.outputSchema, endpoint.port);
      if (!nodeSelection.found) {
        diagnostics.push(
          diagnostic(
            "GE1201_MISSING_SOURCE_PORT",
            `Public output '${outputName}' port '${endpoint.port}' is not a required node output property`,
            `#/outputs/${pointerSegment(outputName)}/port`,
            { outputName, nodeIds: [node.id] },
          ),
        );
      }
    }
    if (nodeSelection.found && !schemaEquals(nodeSelection.schema, graphSelection.schema)) {
      diagnostics.push(
        diagnostic(
          "GE1206_OUTPUT_SCHEMA_MISMATCH",
          `Public output '${outputName}' schema does not exactly match graph outputSchema`,
          `#/outputs/${pointerSegment(outputName)}`,
          { outputName, nodeIds: [node.id] },
        ),
      );
    }
  }

  return Object.freeze(diagnostics);
}

/** Safely validate an untrusted graph value without invoking caller accessors. */
export function validateStrictTypedPorts(
  graph: GraphSpec | unknown,
): readonly CompilerDiagnostic[] {
  const captured = captureTypedGraph(graph);
  if (captured === null) {
    return Object.freeze([
      diagnostic(
        "GE1205_INVALID_PORT_SCHEMA",
        "Typed-port validation requires a portable Graph IR object",
        "#",
      ),
    ]);
  }
  return validateStrictTypedPortsSnapshot(captured);
}

/** Detached value emitted by builder helpers; it does not mutate a graph. */
export const STRICT_TYPED_PORTS_POLICY: StrictTypedPortsPolicy = Object.freeze({
  apiVersion: STRICT_TYPED_PORTS_API_VERSION,
  mode: STRICT_TYPED_PORTS_MODE,
});

// Keep this type reachable for package consumers authoring strict schemas.
export type StrictTypedJsonSchema = JsonSchema;
